lib/puppet/generate/type.rb

```diff
@@ -46,7 +46,24 @@ def format=(format)
         # @return [Boolean] Returns true if the output is up-to-date or false if not.
         def up_to_date?(outputdir)
           f = effective_output_path(outputdir)
-          Puppet::FileSystem::exist?(f) && (Puppet::FileSystem::stat(@path) <=> Puppet::FileSystem::stat(f)) <= 0
+          # Check the fast-path scenarios first.
+          unless Puppet::FileSystem::exist?(f)
+            Puppet.debug("#{f} does not exist.")
```

Comment:
For new code it's preferable to pass a block to `debug` so that we only pay the penalty of constructing the string when debugging is enabled:

```suggestion
            Puppet.debug{"#{f} does not exist."}
```